File: amplify/functions/chat-handler/tools/field-management.ts
import { z } from 'zod';
import {
  FIELD_CONFIGS,
  FIELD_ORDER,
  getFieldDisplayName,
  getFieldEmoji,
} from '../utils/field-configs.js';
import type { DealData } from '../langraph/state.js';

// Port of models/deal_data.py DealData with the same validation constraints.
// camelCase'd; min/max ranges preserved verbatim.
const DealDataSchema = z.object({
  dealName: z.string().min(1),
  purchasePrice: z.number().int().positive(),
  netOperatingIncome: z.number().int().positive(),
  noiGrowthRate: z.number().min(0).max(100),
  holdPeriod: z.number().int().min(1).max(50),
  exitCapRate: z.number().positive().max(100),
  city: z.string().optional(),
});

export function getMissingRequiredFields(
  collected: Record<string, unknown>,
): string[] {
  const missing: string[] = [];
  for (const fieldName of FIELD_ORDER) {
    const cfg = FIELD_CONFIGS[fieldName];
    if (cfg.required && !(fieldName in collected)) {
      missing.push(fieldName);
    }
  }
  return missing;
}

export function getNextFieldToCollect(
  collected: Record<string, unknown>,
): string | null {
  const missingRequired = getMissingRequiredFields(collected);
  if (missingRequired.length > 0) return missingRequired[0];

  for (const fieldName of FIELD_ORDER) {
    const cfg = FIELD_CONFIGS[fieldName];
    if (!cfg.required && !(fieldName in collected)) {
      return fieldName;
    }
  }
  return null;
}

export function getFieldPrompt(fieldName: string): string {
  const cfg = FIELD_CONFIGS[fieldName];
  if (!cfg) {
    const display = getFieldDisplayName(fieldName);
    return `📝 **${display}**\n\nPlease provide the ${display.toLowerCase()}:`;
  }
  const display = getFieldDisplayName(fieldName);
  const emoji = getFieldEmoji(fieldName);
  let out = `${emoji} **${display}**\n\n${cfg.prompt}`;
  if (cfg.examples.length > 0) {
    out += `\n\n*Examples: ${cfg.examples.join(', ')}*`;
  }
  return out;
}

export type ValidationResult =
  | { valid: true; dealData: DealData }
  | { valid: false; error: string };

export function validateAllCollectedFields(
  collected: Record<string, unknown>,
): ValidationResult {
  const missing = getMissingRequiredFields(collected);
  if (missing.length > 0) {
    return {
      valid: false,
      error: `Missing required fields: ${missing.join(', ')}`,
    };
  }
  const parsed = DealDataSchema.safeParse(collected);
  if (parsed.success) {
    return { valid: true, dealData: parsed.data };
  }
  const first = parsed.error.issues[0];
  const fieldName = first.path.join('.');
  return {
    valid: false,
    error: `Validation error for ${fieldName}: ${first.message}`,
  };
}

File: reference-python/field_management.py
# ==============================================================================
# tools/field_management.py
# ==============================================================================

from langchain_core.tools import tool
from typing import Dict, Any, List, Optional, Tuple
from config.field_configs import FIELD_CONFIGS, FIELD_ORDER
from models.deal_data import DealData
from pydantic import ValidationError

@tool
def get_missing_required_fields(collected_fields: Dict[str, Any]) -> List[str]:
    """Determine which required fields are still missing"""
    missing = []
    for field_name in FIELD_ORDER:
        config = FIELD_CONFIGS[field_name]
        if config.required and field_name not in collected_fields:
            missing.append(field_name)
    return missing

@tool
def get_next_field_to_collect(collected_fields: Dict[str, Any]) -> Optional[str]:
    """Get the next field that should be collected"""
    missing_required = get_missing_required_fields.invoke({"collected_fields": collected_fields})
    if missing_required:
        return missing_required[0]
    
    for field_name in FIELD_ORDER:
        config = FIELD_CONFIGS[field_name]
        if not config.required and field_name not in collected_fields:
            return field_name
    return None

@tool
def get_field_prompt(field_name: str) -> str:
    """Get the professionally formatted prompt text for a specific field"""
    if field_name in FIELD_CONFIGS:
        config = FIELD_CONFIGS[field_name]
        field_display = field_name.replace('_', ' ').title()
        
        # Add appropriate emoji based on field type
        emoji = _get_field_emoji(field_name)
        
        # Format the prompt professionally
        prompt = f"{emoji} **{field_display}**\n\n{config.prompt}"
        
        # Add examples if available
        if config.examples:
            prompt += f"\n\n*Examples: {', '.join(config.examples)}*"
            
        return prompt
    
    # Fallback for unknown fields
    field_display = field_name.replace('_', ' ').title()
    return f"📝 **{field_display}**\n\nPlease provide the {field_display.lower()}:"

def _get_field_emoji(field_name: str) -> str:
    """Get appropriate emoji for field type"""
    emoji_map = {
        'deal_name': '📝',
        'purchase_price': '💰',
        'net_operating_income': '📊',
        'noi_growth_rate': '📈',
        'hold_period': '⏰',
        'exit_cap_rate': '🎯',
        'city': '🌆',
        'address': '📍'
    }
    return emoji_map.get(field_name, '📋')

@tool
def validate_all_collected_fields(collected_fields: Dict[str, Any]) -> Tuple[bool, Optional[DealData], str]:
    """Validate that all required fields are collected and create DealData object"""
    missing_required = get_missing_required_fields.invoke({"collected_fields": collected_fields})
    if missing_required:
        return False, None, f"Missing required fields: {', '.join(missing_required)}"
    
    try:
        deal_data = DealData(**collected_fields)
        return True, deal_data, ""
    except ValidationError as e:
        error_details = e.errors()[0]
        field_name = error_details['loc'][0]
        error_msg = error_details['msg']
        return False, None, f"Validation error for {field_name}: {error_msg}"